src/nupic/research/connections.py

```diff
@@ -52,92 +159,165 @@ def __init__(self,
 
     # Save member variables
     self.numCells = numCells
+    assert maxSegmentsPerCell > 0
+    assert maxSynapsesPerSegment > 0
     self.maxSegmentsPerCell = maxSegmentsPerCell
     self.maxSynapsesPerSegment = maxSynapsesPerSegment
 
-    # Mappings
-    self._segments = dict()
-    self._synapses = dict()
+    self._cells = [CellData() for _ in xrange(numCells)]
+    self._synapsesForPresynapticCell = defaultdict(list)
+    self._segmentForFlatIdx = []
+
+    self._numSegments = 0
+    self._numSynapses = 0
+    self._nextFlatIdx = 0
+    self._iteration = 0
+
+
+  def segmentsForCell(self, cell):
+    """ Returns the segments that belong to a cell.
+
+    @param cell (int) Cell index
+
+    @return (set) Segment indices
+    """
 
-    # Indexes into the mappings (for performance)
-    self._segmentsForCell = dict()
-    self._synapsesForSegment = dict()
-    self._synapsesForPresynapticCell = defaultdict(dict)
+    segmentsData = self._cells[cell].segments
+    segments = []
+    for i in xrange(len(segmentsData)):
+      if not segmentsData[i].destroyed:
+        segments.append(Segment(i, cell))
 
-    # Index of the next segment to be created
-    self._nextSegmentIdx = 0
-    # Index of the next synapse to be created
-    self._nextSynapseIdx = 0
+    return segments
 
 
-  def cellForSegment(self, segment):
-    """ Returns the cell that a segment belongs to.
+  def synapsesForSegment(self, segment):
+    """ Returns the synapses on a segment.
 
     @param segment (int) Segment index
 
-    @return (int) Cell index
+    @return (set) Synapse indices
     """
-    return self._segments[segment]
+    segmentData = self.dataForSegment(segment)
+    synapses = []
+    if segmentData.destroyed:
+      raise ValueError("Attempting to access destroyed segment's synapses")
+
+    for i in xrange(len(segmentData.synapses)):
+      if not segmentData.synapses[i].destroyed:
+        synapses.append(Synapse(i, segment))
```

Comment:
Maybe this is already in your list of things to try, but: maybe use list comprehension? I played with it real quick and I think it's faster:

~~~python
    return [Synapse(i, segment)
            for i in xrange(len(segmentData.synapses))
            if not segmentData.synapses[i].destroyed]
~~~